### PR TITLE
Add tuto for custom transforms and custom datapoints in gallery example

### PR DESCRIPTION
**Pull request by NicolasHug**
Bunch of docs.

### CODE REVIEW

File: gallery/plot_datapoints.py [hunk 1]
```diff
@@ -156,41 +174,63 @@ def get_transform(train):
 #    even better to not load the masks at all, but this is not possible in this example, since the bounding boxes are
 #    generated from the masks.
 #
-# How do the datapoints behave inside a computation?
-# --------------------------------------------------
+# .. _datapoint_unwrapping_behaviour:
 #
-# Datapoints look and feel just like regular tensors. Everything that is supported on a plain :class:`torch.Tensor`
-# also works on datapoints.
```
Comment (NicolasHug) [inline gallery/plot_datapoints.py:163]: this wasn't removed, just moved up

File: gallery/plot_datapoints.py [hunk 2]
```diff
@@ -68,27 +91,50 @@
...
 # %%
+# The metadata of ``new_bboxes`` is the same as ``bboxes``, but you could pass
+# it as a parameter to override it. Check the
+# :meth:`~torchvision.datapoints.BoundingBoxes.wrap_like` documentation for
+# more details.
+#
 # Do I have to wrap the output of the datasets myself?
 # ----------------------------------------------------
 #
+# TODO: Move this in another guide - this is user-facing, not dev-facing.
```
Comment (NicolasHug) [inline gallery/plot_datapoints.py:136]: Will address eventually and open an issue so we don't forget. Just LMK if this is OK in principle.

File: docs/source/conf.py [hunk 3]
```diff
@@ -320,7 +320,7 @@ def inject_weight_metadata(app, what, name, obj, options, lines):
       used within the autoclass directive.
     """
 
-    if obj.__name__.endswith(("_Weights", "_QuantizedWeights")):
+    if getattr(obj, ".__name__", "").endswith(("_Weights", "_QuantizedWeights")):
```
Comment (NicolasHug) [inline docs/source/conf.py:323]: you can ignore this. It's because `obj` was sometimes a `property` method coming from the Datapoint classes, and these `property` objects don't have `.__name__`.

File: torchvision/datapoints/_bounding_box.py [hunk 4]
```diff
@@ -42,7 +42,7 @@ class BoundingBoxes(Datapoint):
     canvas_size: Tuple[int, int]
 
     @classmethod
-    def _wrap(cls, tensor: torch.Tensor, *, format: BoundingBoxFormat, canvas_size: Tuple[int, int]) -> BoundingBoxes:
+    def _wrap(cls, tensor: torch.Tensor, *, format: BoundingBoxFormat, canvas_size: Tuple[int, int]) -> BoundingBoxes:  # type: ignore[override]
```
Comment (NicolasHug) [inline torchvision/datapoints/_bounding_box.py:45]: I didn't try to do this "properly", I only tried to stop `mypy` from being a nuisance. If there's a more correct way, feel free to push.

File: torchvision/datapoints/_datapoint.py [hunk 5]
```diff
@@ -25,9 +32,13 @@ def _to_tensor(
             requires_grad = data.requires_grad if isinstance(data, torch.Tensor) else False
         return torch.as_tensor(data, dtype=dtype, device=device).requires_grad_(requires_grad)
 
+    @classmethod
+    def _wrap(cls: Type[D], tensor: torch.Tensor) -> D:
+        return tensor.as_subclass(cls)
```
Comment (NicolasHug) [inline torchvision/datapoints/_datapoint.py:37]: I ended up moving `_wrap` and `wrap_like` in the base class. They were all implemented exactly in the same way for all datapoints except for `BoundingBoxes`, because that one requires meta-data.

I think it's OK to assume that these are a good default implementations for most datapoints, i.e. those that don't require meta-data. What can go wrong anyway?

It makes the example much easier to write and explain, so I think we should do that. I can do it in another PR if you prefer though.

File: torchvision/datapoints/_datapoint.py [hunk 6]
```diff
@@ -25,9 +32,13 @@ def _to_tensor(
...
+
     @classmethod
     def wrap_like(cls: Type[D], other: D, tensor: torch.Tensor) -> D:
-        raise NotImplementedError
+        return cls._wrap(tensor)
```
Comment (NicolasHug) [inline torchvision/datapoints/_datapoint.py:41]: Wait, so now I'm a bit confused. 
We clearly don't use `other` so we should remove it from there.


BUT then:

-  don't we want to allow `wrap_like(other, t)` so that t gets the same device (and dtype??) as `other`??  we don't seem to do that
- do we eve need `wrap_like` for those datapoints that don't have meta-data? What's the point of doing `Image.wrap_like(other, t)` or even `Image.wrap_like(t)` (not supported) when one can just do `Image(t)`?

Anyway, I'd suggest to keep things as-is for now, and to address that stuff after this PR is merged.

Comment (pmeier) [reply torchvision/datapoints/_datapoint.py:41]: The reason we have `wrap_like` the way it is, is that we use it inside `__torch_function__`:

https://github.com/pytorch/vision/blob/9ebf10af02aeb882bd8d7782149e21b48528d562/torchvision/datapoints/_datapoint.py#L32-L39

Basically, we need a generic way for "Here is a datapoint input possibly with some metadata. Given a plain tensor, make it the same type as the datapoint and copy the metadata if available".

So the reason we have `other` in all of them is because if we need metadata and pass nothing explicitly, we need to take it from `other`. This also means that on datapoints that have no metadata, we never use `other`.

Comment (NicolasHug) [reply torchvision/datapoints/_datapoint.py:41]: OK. I understand this is useful for us internally but I'm still questioning whether we should expose that as public for datapoints that don't have meta-data. Let's just keep things as-is in this PR so we can merge it soon, and revisit just after.

File: gallery/plot_datapoints.py [hunk 7]
```diff
@@ -105,8 +151,8 @@ class PennFudanDataset(torch.utils.data.Dataset):
     def __getitem__(self, item):
         ...
 
-        target["boxes"] = datapoints.BoundingBoxes(
-            boxes,
+        target["bboxes"] = datapoints.BoundingBoxes(
+            bboxes,
```
Comment (pmeier) [inline gallery/plot_datapoints.py:155]: I would prefer we stick to the terms given in the tutorial this example is based on: https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html

```suggestion
        target["boxes"] = datapoints.BoundingBoxes(
            boxes,
```

File: gallery/plot_datapoints.py [hunk 8]
```diff
@@ -156,41 +202,65 @@ def get_transform(train):
...
+assert isinstance(new_bboxes, datapoints.BoundingBoxes)
 
 # %%
+# See more details above in :ref:`datapoint_creation`.
+#
+# .. note::
+#
+#    You never need to re-wrap manually if you're using the built-in transforms
+#    or their functional equivalents, because this logic is taken care of for
+#    you.
```
Comment (pmeier) [inline gallery/plot_datapoints.py:238]: Nit: as written, sounds weird to me

```suggestion
#    or their functional equivalents, because this is taken care of for you.
```

File: gallery/plot_datapoints.py [hunk 9]
```diff
@@ -156,41 +202,65 @@ def get_transform(train):
...
-assert isinstance(image, torch.Tensor)
+# image got transformed in-place and is still an Image datapoint, but new_image
+# is a Tensor. They share the same underlying data and they're equal, just
+# different classes.
+assert isinstance(image, datapoints.Image)
 print(image)
 
 assert isinstance(new_image, torch.Tensor) and not isinstance(new_image, datapoints.Image)
 assert (new_image == image).all()
+assert new_image.data_ptr() == image.data_ptr()
```
Comment (pmeier) [inline gallery/plot_datapoints.py:266]: We can do this to hammer the point home, but I feel one of them would be sufficient.

File: torchvision/datapoints/_bounding_box.py (hunk 4 above)
Comment (pmeier) [reply torchvision/datapoints/_bounding_box.py:45]: `mypy` is correct here, but I agree that silencing it is the right move here.

File: torchvision/datapoints/_datapoint.py (hunk 5 above)
Comment (pmeier) [reply torchvision/datapoints/_datapoint.py:37]: I'm ok leaving it here.

File: torchvision/datapoints/_datapoint.py [hunk 10]
```diff
@@ -14,6 +14,13 @@
 
 
 class Datapoint(torch.Tensor):
+    """[Beta] Base class for all datapoints.
+
+    You probably don't want to use this class unless you're defining your own
+    custom Datapoints. See
+    :ref:`sphx_glr_auto_examples_plot_custom_datapoints.py` for details.
```
Comment (pmeier) [inline torchvision/datapoints/_datapoint.py:21]: That sounds a little negative. Can't we just link the gallery for an example?

File: gallery/plot_datapoints.py [hunk 11]
```diff
@@ -156,41 +202,65 @@ def get_transform(train):
...
 # .. note::
 #
 #    This "unwrapping" behaviour is something we're actively seeking feedback on. If you find this surprising or if you
 #    have any suggestions on how to better support your use-cases, please reach out to us via this issue:
 #    https://github.com/pytorch/vision/issues/7319
 #
-# There are two exceptions to this rule:
+# There are two exceptions to this "unwrapping" rule:
 #
 # 1. The operations :meth:`~torch.Tensor.clone`, :meth:`~torch.Tensor.to`, and :meth:`~torch.Tensor.requires_grad_`
```
Comment (pmeier) [inline gallery/plot_datapoints.py:248]: In the meantime we also added `.detach()`

File: torchvision/transforms/v2/functional/_utils.py [hunk 12]
```diff
@@ -47,6 +47,11 @@ def _name_to_dispatcher(name):
 
 
 def register_kernel(dispatcher, datapoint_cls):
+    """Register a kernel for a dispatcher and a (custom) datapoint type.
```
Comment (pmeier) [inline torchvision/transforms/v2/functional/_utils.py:50]: We should at least mention that this is meant as a decorator or provide a minimal code snippet. I know this is explained in the tutorial, but I would prefer to have something here as well.

**Review by pmeier — COMMENTED** [review]
Looking good, thanks Nicolas! I left a bunch of comments.

File: gallery/plot_datapoints.py [hunk 13]
```diff
@@ -68,27 +91,50 @@
...
-# corresponding image alongside the actual values:
+# Some datapoints require additional metadata to be passed in ordered to be constructed. For example,
+# :class:`~torchvision.datapoints.BoundingBoxes` requires the coordinate format as well as the size of the
+# corresponding image (``canvas_size``) alongside the actual values. These
+# metadata are required to properly transform the bounding boxes.
 
-bounding_box = datapoints.BoundingBoxes(
-    [17, 16, 344, 495], format=datapoints.BoundingBoxFormat.XYXY, canvas_size=image.shape[-2:]
+bboxes = datapoints.BoundingBoxes(
+    [[17, 16, 344, 495], [0, 10, 0, 10]], format=datapoints.BoundingBoxFormat.XYXY, canvas_size=image.shape[-2:]
```
Comment (pmeier) [inline gallery/plot_datapoints.py:107]: Improve rendering:

image


```suggestion
    [[17, 16, 344, 495], [0, 10, 0, 10]], 
    format=datapoints.BoundingBoxFormat.XYXY, 
    canvas_size=image.shape[-2:],
```

File: gallery/plot_custom_datapoints.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+_ = t(my_dp)
+
+# %%
+# .. note::
+#
+#     We cannot register a kernel for a transform class, we can only register a
+#     kernel for a **functional**. The reason we can't register a transform
+#     class is because one transform may internally rely on more than one
+#     functional, so in general we can't register a single kernel for a given
+#     class.
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply gallery/plot_custom_datapoints.py:93]: This is a question I asked myself when *writing* the tutorial lol. I think it's a fair question to ask for someone developing their own datapoint, they'll probably wonder "how can I register an operation for this **tranform**" (and not necessarily "for this **dispatcher**").
I also think we can safely have a fair bit of technical details and explanations in this tutorial as it's solely developer-facing. The audience is expected to be technically competent

File: gallery/plot_datapoints.py [hunk 15]
```diff
@@ -68,27 +91,50 @@
...
-# :class:`~torchvision.datapoints.BoundingBoxes` stores the coordinate format as well as the spatial size of the
-# corresponding image alongside the actual values:
+# Some datapoints require additional metadata to be passed in ordered to be constructed. For example,
+# :class:`~torchvision.datapoints.BoundingBoxes` requires the coordinate format as well as the size of the
+# corresponding image (``canvas_size``) alongside the actual values. These
+# metadata are required to properly transform the bounding boxes.
 
-bounding_box = datapoints.BoundingBoxes(
-    [17, 16, 344, 495], format=datapoints.BoundingBoxFormat.XYXY, canvas_size=image.shape[-2:]
+bboxes = datapoints.BoundingBoxes(
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply gallery/plot_datapoints.py:106]: FWIW we already use "bbox" in our own code (in particular in `functional/geometry.py`, but this is user-facing docs so it's a different scope than #7780 and I wouldn't wait on it.

Long lines on those gallery example tend to be harder to read as they get "wrapped around" in the code block and that looks a bit ugly, so I prefer using the shorter `bbox` name here.

(I was even going to `import datapoints as dp` but I knew it would spawn discussions :p so I refrained)

File: gallery/plot_datapoints.py (hunk 7 above)
Comment (NicolasHug) [reply gallery/plot_datapoints.py:155]: I'm not sure, I think we want to be consistent with the names used in this tutorial. What's being used in another tutorial is lower in priority when it comes to consistency

File: torchvision/datapoints/_datapoint.py (hunk 10 above)
Comment (NicolasHug) [reply torchvision/datapoints/_datapoint.py:21]: Happy to consider alternative suggestions but I think we should drive most readers away from this class. My main goal here is that I don't want to scare them and make them believe they need to understand what this is/does just to use v2.

File: torchvision/transforms/v2/functional/_utils.py (hunk 12 above)
Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_utils.py:50]: I had that originally, then I realized this was **exactly** the same code as the one in the gallery example. I don't mind adding it back if you think it's worth it, but it will be redundant.

And since those docstrings examples are never executed (although technically we could have that), we wouldn't be able to catch it if there's ever an issue in that code

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip, I addressed most comments or left some thoughts on the rest. LMK.

File: gallery/plot_custom_transforms.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+# In most cases, this is all you're going to need, as long as you already know
+# the structure of the input that your transform will expect. For example if
+# you're just doing image classification, your transform will typically accept a
+# single image as input, or a ``(img, label)`` input. So you can just hard-code
+# your ``forward`` method to accept just that, e.g.
+#
+# .. code:: python
+#
+#     class MyCustomTransform(torch.nn.Module):
+#         def forward(self, img, label):
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply gallery/plot_custom_transforms.py:35]: Honestly, I don't have an opinion, other than the fact that this probably doesn't matter and I don't feel like re-writing this. If you still think it's better to really illustrate this example the same way we have our datasets structured, I'll re-write.

[the remark this answers is not in the document]
Comment (pmeier) [reply gallery/plot_custom_transforms.py:35]: Leaning towards making this compatible with our datasets. OTOH, datasets aren't even mentioned here so it is a little less of a concern. I'm ok with leaving it as is.

File: torchvision/transforms/v2/functional/_utils.py (hunk 12 above)
Comment (pmeier) [reply torchvision/transforms/v2/functional/_utils.py:50]: Maybe just

```suggestion
    """Decorate a kernel to register it for a dispatcher and a (custom) datapoint type.
```

?

File: gallery/plot_datapoints.py (hunk 7 above)
Comment (NicolasHug) [reply gallery/plot_datapoints.py:157]: OK, I'll move this section into another file, so let's revisit there. I think we should try find a better way to illustrate all this anyway because right now that section isn't even executable

**Review by pmeier — APPROVED** [review]
Thanks Nicolas, awesome to have better docs now!
